### ISSUE TITLE
@ROCDynamicLocalArray: add support for dynamic eltype and expressions for dims

### ISSUE THREAD

**Issue by omlins** [JuliaGPU/AMDGPU.jl#428]
```julia
using AMDGPU

function kernel(C, A, B)
    # Allocate local memory dynamically
    Ctmp = @ROCDynamicLocalArray(eltype(C), (workgroupDim().x, workgroupDim().y))
    # Or, allocate local memory statically if the size is known ahead-of-time
    # Ctmp = @ROCStaticLocalArray(Float64, 8) # if we want 8 elements

    # Use it
    idx = AMDGPU.workitemIdx().x
    Ctmp[idx] = A[idx] + B[idx] + C[1]
    AMDGPU.Device.sync_workgroup()
    C[idx] = Ctmp[idx]
    nothing
end
RA = AMDGPU.rand(4)
RB = AMDGPU.rand(4)
RC = AMDGPU.rand(4)
# Note: The `localmem` option isn't necessary if `@ROCStaticLocalArray` is used
wait(@roc groupsize=4 localmem=sizeof(Float64)*(length(RC)+100) kernel(RC, RA, RB))
```
results in (due to `eltype(C)`):
```julia
ERROR: InvalidIRError: compiling MethodInstance for kernel(::AMDGPU.Device.ROCDeviceVector{Float32, 1}, ::AMDGPU.Device.ROCDeviceVector{Float32, 1}, ::AMDGPU.Device.ROCDeviceVector{Float32, 1}) resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to setindex!)
Stacktrace:
 [1] kernel
   @ ./REPL[63]:9
Reason: unsupported dynamic function invocation (call to getindex)
Stacktrace:
 [1] kernel
   @ ./REPL[63]:11
Reason: unsupported dynamic function invocation (call to setindex!)
Stacktrace:
 [1] kernel
   @ ./REPL[63]:11
Reason: unsupported use of an undefined name (use of 'C')
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:79
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to eltype)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:79
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to alloc_local)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:79
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to ROCDeviceArray)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:79
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported use of an undefined name (use of 'C')
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:84
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to eltype)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:84
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to zero)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:84
 [2] kernel
   @ ./REPL[63]:3
Reason: unsupported dynamic function invocation (call to setindex!)
Stacktrace:
 [1] macro expansion
   @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/device/gcn/memory_static.jl:84
 [2] kernel
   @ ./REPL[63]:3
Hint: catch this exception as `err` and call `code_typed(err; interactive = true)` to introspect the erronous code with Cthulhu.jl
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}, args::LLVM.Module)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/validation.jl:154
  [2] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:407 [inlined]
  [3] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/TimerOutputs/RsWnF/src/TimerOutput.jl:253 [inlined]
  [4] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:406 [inlined]
  [5] emit_llvm(job::GPUCompiler.CompilerJob; libraries::Bool, deferred_codegen::Bool, optimize::Bool, cleanup::Bool, only_entry::Bool, validate::Bool, ctx::LLVM.ThreadSafeContext)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/utils.jl:89
  [6] codegen(output::Symbol, job::GPUCompiler.CompilerJob; libraries::Bool, deferred_codegen::Bool, optimize::Bool, cleanup::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing, ctx::LLVM.ThreadSafeContext)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:120
  [7] compile(target::Symbol, job::GPUCompiler.CompilerJob; libraries::Bool, deferred_codegen::Bool, optimize::Bool, cleanup::Bool, strip::Bool, validate::Bool, only_entry::Bool, ctx::LLVM.ThreadSafeContext)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:90
  [8] (::AMDGPU.Compiler.var"#61#63"{GPUCompiler.CompilerJob{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}})(ctx::LLVM.ThreadSafeContext)
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:125
  [9] LLVM.ThreadSafeContext(f::AMDGPU.Compiler.var"#61#63"{GPUCompiler.CompilerJob{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}})
    @ LLVM /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/LLVM/bsdku/src/executionengine/ts_module.jl:14
 [10] JuliaContext
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:35 [inlined]
 [11] compile
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:124 [inlined]
 [12] actual_compilation(cache::Dict{UInt64, Any}, key::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}, ft::Type, tt::Type, compiler::typeof(AMDGPU.Compiler.compile), linker::typeof(AMDGPU.Compiler.link))
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/cache.jl:53
 [13] cached_compilation(cache::Dict{UInt64, Any}, cfg::GPUCompiler.CompilerConfig{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}, ft::Type, tt::Type, compiler::Function, linker::Function)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/cache.jl:37
 [14] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:111 [inlined]
 [15] macro expansion
    @ ./lock.jl:267 [inlined]
 [16] rocfunction(f::typeof(kernel), tt::Type; device::ROCDevice, global_hooks::NamedTuple{(), Tuple{}}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:105
 [17] rocfunction(f::typeof(kernel), tt::Type)
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:101
 [18] top-level scope
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/highlevel.jl:599

```

And the following
```julia
using AMDGPU

function kernel(C, A, B)
    # Allocate local memory dynamically
    Ctmp = @ROCDynamicLocalArray(Float64, (workgroupDim().x, workgroupDim().y+2))
    # Or, allocate local memory statically if the size is known ahead-of-time
    # Ctmp = @ROCStaticLocalArray(Float64, 8) # if we want 8 elements

    # Use it
    idx = AMDGPU.workitemIdx().x
    Ctmp[idx] = A[idx] + B[idx] + C[1]
    AMDGPU.Device.sync_workgroup()
    C[idx] = Ctmp[idx]
    nothing
end
RA = AMDGPU.rand(4)
RB = AMDGPU.rand(4)
RC = AMDGPU.rand(4)
# Note: The `localmem` option isn't necessary if `@ROCStaticLocalArray` is used
wait(@roc groupsize=4 localmem=sizeof(Float64)*(length(RC)+100) kernel(RC, RA, RB))
```
results in (due to `workgroupDim().y+2`):
```julia
ERROR: GPU compilation of MethodInstance for kernel(::AMDGPU.Device.ROCDeviceVector{Float32, 1}, ::AMDGPU.Device.ROCDeviceVector{Float32, 1}, ::AMDGPU.Device.ROCDeviceVector{Float32, 1}) failed
KernelError: kernel returns a value of type `Union{}`

Make sure your kernel function ends in `return`, `return nothing` or `nothing`.
If the returned value is of type `Union{}`, your Julia code probably throws an exception.
Inspect the code with `@device_code_warntype` for more details.

Stacktrace:
  [1] check_method(job::GPUCompiler.CompilerJob)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/validation.jl:41
  [2] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:100 [inlined]
  [3] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/TimerOutputs/RsWnF/src/TimerOutput.jl:253 [inlined]
  [4] codegen(output::Symbol, job::GPUCompiler.CompilerJob; libraries::Bool, deferred_codegen::Bool, optimize::Bool, cleanup::Bool, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing, ctx::LLVM.ThreadSafeContext)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:99
  [5] compile(target::Symbol, job::GPUCompiler.CompilerJob; libraries::Bool, deferred_codegen::Bool, optimize::Bool, cleanup::Bool, strip::Bool, validate::Bool, only_entry::Bool, ctx::LLVM.ThreadSafeContext)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:90
  [6] (::AMDGPU.Compiler.var"#61#63"{GPUCompiler.CompilerJob{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}})(ctx::LLVM.ThreadSafeContext)
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:125
  [7] LLVM.ThreadSafeContext(f::AMDGPU.Compiler.var"#61#63"{GPUCompiler.CompilerJob{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}})
    @ LLVM /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/LLVM/bsdku/src/executionengine/ts_module.jl:14
  [8] JuliaContext
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/driver.jl:35 [inlined]
  [9] compile
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:124 [inlined]
 [10] actual_compilation(cache::Dict{UInt64, Any}, key::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}, ft::Type, tt::Type, compiler::typeof(AMDGPU.Compiler.compile), linker::typeof(AMDGPU.Compiler.link))
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/cache.jl:53
 [11] cached_compilation(cache::Dict{UInt64, Any}, cfg::GPUCompiler.CompilerConfig{GPUCompiler.GCNCompilerTarget, AMDGPU.Compiler.ROCCompilerParams}, ft::Type, tt::Type, compiler::Function, linker::Function)
    @ GPUCompiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/GPUCompiler/cy24l/src/cache.jl:37
 [12] macro expansion
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:111 [inlined]
 [13] macro expansion
    @ ./lock.jl:267 [inlined]
 [14] rocfunction(f::typeof(kernel), tt::Type; device::ROCDevice, global_hooks::NamedTuple{(), Tuple{}}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:105
 [15] rocfunction(f::typeof(kernel), tt::Type)
    @ AMDGPU.Compiler /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/compiler/codegen.jl:101
 [16] top-level scope
    @ /scratch/project_465000105/saomlin/julia_local/julia_depot/packages/AMDGPU/9VGcR/src/highlevel.jl:599
```

**Comment by omlins** [comment]
@jpsamaroo : there might be some missing `esc` in this macro?

**Comment by jpsamaroo** [comment]
Former example fixed in #430

The latter fails because `ROCDeviceArray` expects `dims` as a `Tuple` of the same type of `Integer`; in the above case, they're in a `Tuple{UInt32, Int64}`, hence the error. You can do `workgroupDim().y+0x2` to keep it from promoting.
